### DISCUSSION TITLE
Dynamically refreshed DB Credentials

### DISCUSSION THREAD

**Discussion by zikphil** [sqlalchemy/sqlalchemy#8349] [Usage Questions] [answered]
Hello fantastic community,

I use Vault to generate 90m long db credentials. When my Python API starts, I call this:

```
    bind = create_async_engine(
        f"postgresql+asyncpg://?requested_db=foobar''}",
        pool_recycle=3600
    )
```

And I have a custom event listener:

```
    @event.listens_for(Engine, "do_connect")
    def dynamically_provide_engine_credentials(dialect, conn_rec, cargs, cparams):
        requested_db = cparams.get('requested_db', False)

        if requested_db:

            print('getting new credentials')

            # Using the Application Dynamic Configuration, we pull credentials for this DB and pass it to the engine.
            db_credentials = GET CREDENTIALS FROM VAULT

            del cparams['requested_db']
            cparams['host'] = db_credentials['host']
            cparams['port'] = db_credentials['port']
            cparams['user'] = db_credentials['username']
            cparams['password'] = db_credentials['password']
            cparams['database'] = db_credentials['db_name']
```

So this is working fine for the initial connect, it gets the custom credentials from vault and set ups the connection properly.

But I was under the impression that by using a `pool_recycle=3600` SQLAlchemy would kill the connection after 3600 and this event listener would be called again to get credentials to setup a new connection, but it is not the case. The Pool is requesting a new connection using the same credentials as before without triggering this event listener.

Am I listening to the wrong event?

**Comment by zikphil** [comment, marked as answer]
Actually nevermind.

My problem is that I am deleting `requested_db` so its not there for the next run of the event.

**Reply by CaselIT** [reply]
Also note that sqlalchemy will close a connection only if it's in the queue. If you checkout a connection, do stuff for 2h then continue to use it that connection is still on the old credentials since no one closed it. 
The first time it will get back to the queue it will be closed

**Reply by zikphil** [reply]
@CaselIT If I have a sessionmaker in my API server and that for every request I generate a new session and close it at the end of the request I shouldnt encounter this problem no?

**Reply by CaselIT** [reply]
No, I don't think so, since the connection would be checked out only for the duration of the request (assuming that a request lasts less than the recycle time)

**Reply by zikphil** [reply]
@CaselIT Well as it turns out, I must be having some leaking sessions not being closed properly because my app is experiencing issues with SQLAlchemy using connections that are still on old credentials and therefor stop working after a while. Is there another way for me to implement somewhat of a safety check that would be non-destructive?

Like... When a connection is used it would do a manual ttl check and if it's past the pool_recycle, it would kill itself and generate a fresh one for the query being requested?

I have been trying to figure this out for days but it's very hard to troubleshoot which connections are alive at which point and what gets used when. I think figuring out a "master" way to always recycle connections, used or not, after ttl would allow me more time.

any thoughts on the best way of doing this?

**Reply by CaselIT** [reply]
I don't think there is anything build it, but maybe you can do something by using one of the events to invalidate the connection?

You can access the creation time of the connection using private api as `conn.connection._connection_record.last_connect_time` where `conn` is an engine connection, or in a pool event that has the a `connection_record` like [checkout](https://docs.sqlalchemy.org/en/14/core/events.html?highlight=event#sqlalchemy.events.PoolEvents.checkout) doing `connection_record.last_connect_time`

In any case i would suggest ensuring that sessions are not kept alive. If you are using a scoped session you should remember to call `remove` once done with a particular scope

**Reply by zikphil** [reply]
@zzzeek After more investigation I believe you are correct to assume that I have some sessions that never gets closed and are then re-used later down the line when they shouldnt. What would be a good way for me to troubleshoot that? Is there a way to access a global registry of all opened sessions in my app? I tried accessing `_sessions` from orm.session but it is not exported.

**Reply by zzzeek** [reply]
you should be able to look in _sessions, that is in fact all the current Session objects in a weak referencing dictionary.  `from sqlalchemy.orm.session import _sessions`
